src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexPrefixAnalyzer.cs

```diff
@@ -514,16 +487,197 @@ static bool TryFindFixedSets(RegexNode node, List<(char[]? Chars, string Set, in
         /// </summary>
         public static string? FindFirstCharClass(RegexNode root)
         {
-            var s = new RegexPrefixAnalyzer(stackalloc int[StackBufferSize]);
-            RegexFC? fc = s.RegexFCFromRegexTree(root);
-            s.Dispose();
-
-            if (fc == null || fc._nullable)
+            // Explore the graph, adding found chars into a result set, which is lazily initialized so that
+            // we can initialize it to a parsed set if we discover one first (this is helpful not just for allocation
+            // but because it enables supporting starting negated sets, which wouldn't work if they had to be merged
+            // into a non-negated default set). If the operation returns true, we successfully explore all relevant nodes
+            // in the graph.  If it returns false, we were unable to successfully explore all relevant nodes, typically
+            // due to conflicts when trying to add characters into the result set, e.g. we may have read a negated set
+            // and were then unable to merge into that a subsequent non-negated set.  If it returns null, it means the
+            // whole pattern was nullable such that it could match an empty string, in which case we
+            // can't make any statements about what begins a match.
+            RegexCharClass? cc = null;
+            return TryFindFirstCharClass(root, ref cc) == true ?
+                cc!.ToStringClass() :
+                null;
+
+            // Walks the nodes of the expression looking for any node that could possibly match the first
+            // character of a match, e.g. in `a*b*c+d`, we'd find [abc], or in `(abc|d*e)?[fgh]`, we'd find
+            // [adefgh].  The function is called for each node, recurring into children where appropriate,
+            // and returns:
+            // - true if the child was successfully processed and represents a stopping point, e.g. a single
+            //   char loop with a minimum bound greater than 0 such that nothing after that node in a
+            //   concatenation could possibly match the first character.
+            // - false if the child failed to be processed but needed to be, such that the results can't
+            //   be trusted.  If any node returns false, the whole operation fails.
+            // - null if the child was successfully processed but doesn't represent a stopping point, i.e.
+            //   it's zero-width (e.g. empty, a lookaround, an anchor, etc.) or it could be zero-width
+            //   (e.g. a loop with a min bound of 0).  A concatenation processing a child that returns
+            //   null needs to keep processing the next child.
+            static bool? TryFindFirstCharClass(RegexNode node, ref RegexCharClass? cc)
```

Comment:
NIT: should cc parameter have the attribute `NotNullWhen(true)` so that you don't have to use the `!` operator at the callsite above?

Reply:
`NotNullWhen(true)` only applies to `bool`.  It unfortunately doesn't work with `bool?`.
Related: https://github.com/dotnet/roslyn/issues/57649